File: 0328_odd_even_linked_list/odd_even_list.py
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def odd_even_list(head: ListNode) -> ListNode:
    """
    Time: O(n)
    Space: O(1)
    """
    odd = odd_head = ListNode(-1)
    even = even_head = ListNode(-1)
    while head:
        odd.next = head
        odd = odd.next
        even.next = head.next
        even = even.next
        head = head.next.next if head.next else None
    odd.next = even_head.next
    return odd_head.next

File: 0328_odd_even_linked_list/odd_even_list.go
package main

type ListNode struct {
	Val  int
	Next *ListNode
}

func oddEvenList(head *ListNode) *ListNode {
	odd := &ListNode{-1, nil}
	oddHead := odd
	even := &ListNode{-1, nil}
	evenHead := even
	for head != nil {
		odd.Next = head
		odd = odd.Next
		even.Next = head.Next
		even = even.Next
		if head.Next != nil {
			head = head.Next.Next
		} else {
			head = nil
		}
	}
	odd.Next = evenHead.Next
	return oddHead.Next
}

func main() {

}
